x/programs/rust/examples/counter/src/lib.rs

```diff
@@ -45,7 +45,7 @@ pub fn inc(context: Context, to: Address, amount: i64) -> bool {
 #[public]
 pub fn inc_external(_: Context, target: Program, max_units: i64, of: Address, amount: i64) -> i64 {
     let params = params!(&of, &amount).unwrap();
```

Comment:
What would you change ?

Reply:
If you look at the `params!` definition here:
https://github.com/ava-labs/hypersdk/blob/3fdf8bfdddd50496e45956efb3de1ca4aa0edb61/x/programs/rust/wasmlanche-sdk/src/params.rs#L7-L13

You can see that each param is serialized eagerly, then all the serialized params are put into a `Vec<Param>`, which is then also serialized into `Params(Vec<u8>)`. But with the changes in this PR, we serialized this _again_  ... it's super inefficient. It kind of made sense before, but it definitely doesn't make sense to do this now. 

I'm thinking that `Params(Vec<u8>)` -> `Vec<Box<dyn BorshSerialize>>` , and serialize it in one shot here:
https://github.com/ava-labs/hypersdk/pull/881/files/3fdf8bfdddd50496e45956efb3de1ca4aa0edb61#diff-df55dbd18bc9cea6bdf585205a2306b353f57c2a679a4e6c735a64d043aed93eR63

Make sense? Can clarify further if you need.

Reply:
That does make sense, thanks for the clarification ! 
I think that it would be a sensible change, and after looking at the `serialize_param` it's doing exactly what Borsh does when treating dynamically sized arrays, if you look at the specification on https://borsh.io/
https://github.com/ava-labs/hypersdk/blob/f002156f55e422c79631904582f77ed0811d7419/x/programs/rust/wasmlanche-sdk/src/params.rs#L46-L51

edit: that won't work because `borsh::to_vec` takes as an input a generic which implements `BorshSerialize + ?Sized`

Reply:
First of all, I suggested something that wouldn't work; `BorshSerialize` isn't object-safe because its only method is generic. There might be some kind of workaround though. In any case, I'll create an issue.

Reply:
New issue: https://github.com/ava-labs/hypersdk/issues/885